### DISCUSSION TITLE
getStaticProps with Custom App

### DISCUSSION THREAD

**Discussion by andreas1327250** [vercel/next.js#11334] [Help]
getInitialProps in _app.js is also called build-time when using getStaticProps - but how can Props from getStaticProps ever be passed to the Page? The custom app can only render the Page with the Props received from the getInitialProps of _app.js and the getInitialProps of the Page itself.

What do i miss here?

**Comment by MickCoelho** [comment]
See https://github.com/zeit/next.js/discussions/10949

**Reply by andreas1327250** [reply]
Not sure if this is really the same - I don't need a global getStaticProps in _app.js, I just want one page to make use of getStaticProps, while still having my custom _app.js. Basically it seems that the Props from getStaticProps are not usable currently if you have a custom _app.js, is that correct?

**Reply by andreas1327250** [reply]
Seems like other people are having the same problem with understanding all of this:
https://github.com/zeit/next.js/discussions/11192
https://github.com/zeit/next.js/discussions/11183

Seems like a custom _app.js is completely incompatible with getStaticProps and getServerSideProps - though that's not really mentioned anywhere in the docs :/

**Reply by MickCoelho** [reply]
> Not sure if this is really the same - I don't need a global getStaticProps in _app.js, I just want one page to make use of getStaticProps, while still having my custom _app.js. Basically it seems that the Props from getStaticProps are not usable currently if you have a custom _app.js, is that correct?

It seems to work fine for me. On my current project, every single page has a `getStaticProps` while the `_app.js` has a `getInitialProps`.

**Reply by andreas1327250** [reply]
🤔 But how.. if you render each page manually in _app.js with the Props from getInitialProps:
`return <Component {...pageProps} />`

**Reply by MickCoelho** [reply]
Inside the `_app.js` `getInitialProps`, you do your thing and return the data like that:
```
//fetching data
...

return { appProps };
```
then, when you render the page, you pass it down like that: 
```
<Component {...appProps} {...pageProps} />
```

**Reply by andreas1327250** [reply]
"doing my thing" would be calling getStaticProps of that Page. I don't see a reference to the method of the page anywhere.

**Reply by MickCoelho** [reply]
All the code above goes into the `_app.js`.
In your page, you'd have something like that: 
```
export async function getStaticProps() {
 // Fetch your data
  return {
    props: {
      pageContent,
    },
  }
}
```

**Reply by andreas1327250** [reply]
Thanks for the help, but that's exactly the problem. The props (pageContent) you return from here never get into the Page because of the manual rendering of the Page in _app.js with `<Component {...appProps} {...pageProps} />`

**Reply by MickCoelho** [reply]
It works fine on my current project (using nextjs `9.3.0`). Don't you have something inside either `appProps` or `pageProps` named `pageContent` that would override whatever you do inside your page's `getStaticProps`?

**Reply by andreas1327250** [reply]
So you are saying that Next-JS magically merges the manual `...appProps` and `...pageProps` with the returned props from getStaticProps? 🤔 Would be cool, but I don't get what I am doing wrong. Thanks anyway 👍🙇

**Reply by MickCoelho** [reply]
`appProps` and `pageProps` are props sent down to the page itself. Both of those being deconstructed when sent to the page, I wouldn't be surprised that whatever props you store in the page's `getStaticProps` is being overwritten. 
Maybe just try the following in your page, see if that get's down in the render: 
```
export async function getStaticProps() {
 // Fetch your data
  return {
    props: {
      bar: 'foo',
    },
  }
}
```

**Reply by andreas1327250** [reply]
The props I receive in the Page are exactly those which get passed in _app.js. No magic insertion of the getStaticProps-prop happening for me.

**Reply by andreas1327250** [reply]
@timneutkens Can you clarify this maybe, I am completely lost :/ thanks.

**Reply by clfhhc** [reply]
I read through the codes and documents for hours and finally got it working.

pages/_app.tsx:
```tsx
import React from 'react';
import { NextComponentType } from 'next';
import { AppContext, AppProps } from 'next/app';

export interface ModifiedAppInitialProps<A = { [key in string]: string }> {
  appProps: A;
}

export interface ExtendedAppProps<P = { [key in string]: string }, A = { [key in string]: string }>
  extends AppProps<P>,
    ModifiedAppInitialProps<A> {}

const myApp: NextComponentType<AppContext, ModifiedAppInitialProps, ExtendedAppProps> = ({
  Component,
  pageProps,
  appProps,
}) => {
  return <Component {...appProps} {...pageProps} />;
};

myApp.getInitialProps = async () => {
  return { appProps: { /* ...someAppProps */ };
};

export default myApp;
```

pages/index/[id].tsx:
```tsx
import React from 'react';
import { GetStaticProps, GetStaticPaths, NextPage } from 'next';
import { useRouter } from 'next/router';

export interface Props {
  id?: string;
}

const IndexPage: NextPage<Props> = ({ id, ...appProps }) => {
  console.log(appProps);
  const router = useRouter();
  return (
    <div>
      Param:
      {router.isFallback ? 'Hold on' : id}
    </div>
  );
};

export const getStaticProps: GetStaticProps<Props> = async ({ params, ...otherProps }) => {
  return { props: { ...params, ...otherProps } } as { props: Props };
};

export const getStaticPaths: GetStaticPaths = async () => ({
  paths: [{ params: { id: '1' } }],
  fallback: true,
});
export default IndexPage;
```

We have to separate `appProps` and `pageProps` so that the `appProps` doesn't get overwritten: https://github.com/zeit/next.js/blob/master/packages/next/next-server/server/render.tsx#L626

notes: Both `appProps` and `pageProps` need to be serializable.

**Reply by knieber** [reply]
@clfhhc thanks for this example. If I were to perform some sort of external data fetch in `App.getInitialProps`, will that fetch still run at build time and produce a statically generated IndexPage? Or can I only have access to those `appProps` from getInitialProps at build time with a synchronous operation?

**Reply by iamkumaran** [reply]
@clfhhc `appProps` works for me. Thanks for this.

**Reply by benjaminpreiss** [reply]
@clfhhc Having the same problem but sadly I don't know typescript :( Thanks for the solution though!

**Reply by clfhhc** [reply]
> @clfhhc thanks for this example. If I were to perform some sort of external data fetch in `App.getInitialProps`, will that fetch still run at build time and produce a statically generated IndexPage? Or can I only have access to those `appProps` from getInitialProps at build time with a synchronous operation?

sorry, I just saw this. Probably you have the answer already. `App.getInitialProps` is runtime and `getStaticProps` is build time.

**Comment by he4amm** [comment]
I have the same issue as yours in my custom app, any hope fixing this?

**Reply by andreas1327250** [reply]
Ok, I'm further down the rabbit-hole.

Next injects the static props into the "pageProps" of the render-function of _app.js. But if you have a getInitialProps-function in it and also return { pageProps } in there, it overrides them. I am now thinking of putting a flag on the Page-Component (PageComponent.hasStaticProps = true), checking that flag in getInitialProps of _app.js and not returning { pageProps } from there if true.

**Reply by joshuaellis** [reply]
I had this problem but upgraded to v9.3.2 and it now works, I recommend you try that if you haven't already.

**Comment by e-karma** [comment]
I think we could all gain tremendously by additional documentation and an example of fully dynamic routing working.  I have spent a couple days trying to find answers and get this to work.  I now have things working, with one caveat.  The home page, (no route) yields an error "An unexpected error has occurred." in the body of the page only, but no other pages.  

-My API Page data is loaded in _app.js and thats where the global nav lives.
-The Universal page template is src/pages/[urlSlug].js

**Comment by aamorozov** [comment]
Is there a way one could use getStaticProps instead of getInitialProps in _document component?

**Reply by tkstang** [reply]
@aamorozov FWIW I don't think getInitialProps is actually doing the same thing in _document as it would in _app or any other pages.  I recall seeing one of the maintainers stating as much in another thread and that they probably should have named it something else so as I understand it there is no reason to make any changes to getInitialProps in _document.

**Reply by vihapr** [reply]
not sure about using it in _document, but it's exacly not operable in the _app.js component (v9.5.5)

**Comment by alex-grover** [comment]
may or may not be related - it seems like a `Page.getStaticProps` is injected into the App's `pageProps` _after_ `App.getInitialProps` runs, not before as the documentation indicates. For me, pageProps was always an empty object in my `App.getInitialProps` until I changed to using `Page.getInitialProps`.

**Comment by kristojorg** [comment]
Hi all, please let me know what you think of this proposal to add app-wide and path-wide static data fetching: https://github.com/vercel/next.js/discussions/18052
